File: 0998-maximum-binary-tree-ii/0998-maximum-binary-tree-ii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* insertIntoMaxTree(TreeNode* root, int val) {
        //Base Case
        if(!root) return new TreeNode(val);
        
        //if val is greater than the current root.val then it should be attached to right of this new root.
        if(root->val<val) return new TreeNode(val,root,nullptr);
        //recursive call to find the real position of val if the about if condition does not apply.
        root->right = insertIntoMaxTree(root->right,val);
        //just return our newly modified trees root.
        return root;
    }
};

File: 0998-maximum-binary-tree-ii/0998-maximum-binary-tree-ii.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode insertIntoMaxTree(TreeNode root, int val) {
        //Base case if only val exists in b list.
        if(root == null) return new TreeNode(val);
        //when val is greater than the current subtree or whole trees root;
        if(root.val<val) return new TreeNode(val,root,null);        
        //recursive call to search the righjt position of val.
        root.right = insertIntoMaxTree(root.right,val);
        
        return root;
    }
}